staging/src/k8s.io/api/admissionregistration/v1alpha1/types.go

```diff
@@ -419,6 +431,59 @@ type ValidatingAdmissionPolicyBindingSpec struct {
 	// Required.
 	// +listType=set
 	ValidationActions []ValidationAction `json:"validationActions,omitempty" protobuf:"bytes,4,rep,name=validationActions"`
+
+	// NamespaceParamRef specifies the parameter resource or resources used to
+	// configure the admission control policy. It should point to a resources of
+	// the type specified in ParamKind of the bound ValidatingAdmissionPolicy.
+	//
+	// If the policy specifies a ParamKind and the resource referred to by
+	// ParamRef does not exist, this binding is considered mis-configured and
+	// the FailurePolicy of the ValidatingAdmissionPolicy is applied.
+	//
+	// if the paramKind of the policy referred to by `policyName` is cluster scoped,
+	// and namespaceParamRef set, the binding is considered mis-configured, and
+	// the failureMode applies.
+	//
+	// `namespaceParamRef` and `paramRef` are members of a union; if one of the
+	// fields is set, the other must be unset.
+	// +optional
+	NamespaceParamRef *NamespaceParamRef `json:"namespaceParamRef,omitempty" protobuf:"bytes,5,opt,name=namespaceParamRef"`
+}
+
+// NamespaceParamRef references a parameter resource scoped to the same
+// namespace as the resource being evaluated by a policy.
+type NamespaceParamRef struct {
+	// Name is the name of the namespace-scoped param to evaluate against.
+	// When evaluating the admission of a resource request, a param with the
+	// given name will be found in the same namespace as the resource and used
+	// as the `param` input to the policy's rules.
+	//
+	// Exactly one of `name` or `selector` must be specified, but not both.
+	//
+	// +optional
+	Name string `json:"name" protobuf:"bytes,1,opt,name=name"`
+
+	// Selector can be used to match param resources based on its metadata.
+	// If multiple objects are found to match, the policy will be evaluated
+	// against ALL of them and the result will be ANDed together.
+	// If no objects match the selector, then the `failAction` applies.
+	//
+	// Exactly one of `name` or `selector` must be specified, but not both.
+	//
+	// +optional
+	Selector *metav1.LabelSelector `json:"selector" protobuf:"bytes,2,opt,name=selector"`
+
+	// ParameterNotFoundAction controls the behavior of the binding when the resource
+	// exists, and name or selector is valid, but there are no parameters
+	// matched by the binding. If the value is set to `Allow`, then no
+	// matched parameters will be treated as successful validation by the binding.
+	// If set to `Deny`, then no matched parameters will be subject to the
+	// `failurePolicy` of the policy.
+	//
+	// Allowed values are `Allow` or `Deny`
+	// Default to `Allow`
+	// +optional
+	ParameterNotFoundAction *ParameterNotFoundActionType `json:"parameterNotFoundAction" protobuf:"bytes,3,opt,name=parameterNotFoundAction,casttype=ParameterNotFoundActionType"`
```

Comment:
For context: your proposal from your other comment

> If paramKind is cluster-scoped, namespace being set is considered an error
>
> If paramKind is namespaced:
>
> when namespace is blank, it is looked up in the request namespace
> when namespace is set, it is looked up in the named namespace

I think @deads2k or @cici37 have context on the discussion before the design was decided upon in the KEP.

Reply:
I initially only apply selector/parameterNotFoundAction to namespaceParamRef because 
1. for cluster scoped param, ParamRef w/ Allow Action does not make sense since we assume whenever policy.paramKind and binding.paramRef set together, it should point to a valid param resource. Otherwise should err out.
2. for cluster scoped param, the selector under binding.MatchResources could used same ad selector here.
3. for namespacedParam, we wanna it to be valid for both the namespace which the corresponding param resource does not exist  and the namespace where the corresponding param resource exists(ParamRef w/ Allow Action will allow the first case be valid)

I am ok with adding selector into cluster scoped case but concerning about applying the parameterNotFoundAction to cluster scoped will lead to unnecessary behavior change.

Reply:
> 2. for cluster scoped param, the selector under binding.MatchResources could used same ad selector here

I'm not sure I understand how `binding.MatchResources` applies here... isn't the selector in `namespacedParam` related to selecting the _param_ objects? I thought the use case there was selecting a set of param objects by label rather than by name, and was trying to understand why that use case only applied to namespaced param objects and not cluster-scoped param objects.

> 3. for namespacedParam, we wanna it to be valid for both the namespace which the corresponding param resource does not exist and the namespace where the corresponding param resource exists(ParamRef w/ Allow Action will allow the first case be valid)

does this assume the policy rules are written in a way to tolerate `params` being `nil` and evaluate them that way, or skip policy evaluation entirely?

> I am ok with adding selector into cluster scoped case but concerning about applying the parameterNotFoundAction to cluster scoped will lead to unnecessary behavior change.

It seemed like it might be simpler to implement and understand if we unified the behavior of single-instance cluster-scoped params, single-instance namespace-scoped params, and multi-instance per-namespace params. What do you think?

Reply:
Unified the behavior makes sense. Just to confirm the understanding aligned:
We are moving selector and ParameterNotFoundAction upper to under BindingParamRef instead of namespacedParamRef. The expected behavior would be:
- No paramKind -> Validate with null param
- ParamKind set on policy, but no paramRef/namespaceParamRef on binding -> Validate with null param
- NotFound when using ClusterWideParam w/ Deny Action -> error
- NotFound when using ClusterWideParam w/ Allow Action -> skip/pass binding
- NotFound when using NamespaceParamRef w/ Deny Action -> error
- NotFound when using NamespaceParamRef w/ Allow Action -> skip/pass binding

Thank you!

Reply:
I think that makes sense, and means we only need a single paramRef field.

Are the following combinations what you expect:

* cluster-scoped paramKind, paramRef sets `namespace`: error
* cluster-scoped paramKind, paramRef does not set `namespace`: resolve cluster-scoped param objects normally
* namespace-scoped paramKind, paramRef sets `namespace`: resolve param object(s) in the specified namespace
* namespace-scoped paramKind, paramRef does not set `namespace`, request is namespace-scoped: resolve param object(s) in the request's namespace
* namespace-scoped paramKind, paramRef does not set `namespace`, request is cluster-scoped: error or Validate with null param?

Reply:
Does the following type+doc capture this discussion well?

```go
// ParamRef describes how to locate the params to be used as input to
// expressions of rules applied by a policy binding.
//
// Parameters may be identified by name and selector.
//
// If name and `selector` are both specified, then admission will be evaluated
// on params with the provded `name` OR `selector`. If multiple are found, they
// are all evalauted and the results are ANDed together.
//
// A single parameter used for all admission requests can be configurated
// by setting the `name` field, leaving `selector` blank, and setting namespace
// if `paramKind` is namespace-scoped.
//
// A per-namespace parameter may be used by specifying a namespace-scoped
// `paramKind` in the policy, setting `name` or `selector` and leaving `namespace`
// in `paramRef` empty.
//
// +union
type ParamRef struct {
	// Name of the resource being referenced.
	//
	// +optional
	Name string

	// Selector can be used to match multiple param objects based on their labels.
	//
	// +optional
	Selector *metav1.LabelSelector

	// Namespace of the referenced resource. Allows limiting the search for params
	// to a specific namespace. Applies to both `name` and `selector` fields.
	//
	// - If `paramKind` is cluster-scoped, this field MUST be unset. Setting this
	// field results in a configuration error.
	// - If `paramKind` is namespace-scoped, the namespace of the object being
        // evaluated for admission will be used.
	//
	// +optional
	Namespace string

	// ParameterNotFoundAction controls the behavior of the binding when the resource
	// exists, and name or selector is valid, but there are no parameters
	// matched by the binding. If the value is set to `Allow`, then no
	// matched parameters will be treated as successful validation by the binding.
	// If set to `Deny`, then no matched parameters will be subject to the
	// `failurePolicy` of the policy.
	//
	// Allowed values are `Allow` or `Deny`
	// Default to `Allow`
	// +optional
	ParameterNotFoundAction *ParameterNotFoundActionType
}
```

Notably this removes the `paramKind` field in favor of inferring its value from whether `namespace` is set or unset.

Reply:
looks mostly like what I expect, with a couple exceptions:

```
// If name and `selector` are both specified, then admission will be evaluated
// on params with the provded `name` OR `selector`. If multiple matches are found
// all will be evaluated against the policy.
```

that's surprising... I expected name and selector to be mutually exclusive

```
	// - If `paramKind` is namespace-scoped, the namespace of the object being
        // evaluated for admission will be used when this field is left unset.
```
specify what happens if `paramKind` is namespace-scoped, the paramRef namespace is not specified, and the object being evaluated for admission is not namespaced

Reply:
```
	// Allowed values are `Allow` or `Deny`
	// Default to `Allow`
	// +optional
	ParameterNotFoundAction *ParameterNotFoundActionType
```

(I'd probably also recommend dropping this default and making this be explicitly specified in the next alpha or beta version of the API)

Reply:
+1 to mutually exclusive on selector/name. For ParameterNotFoundAction, I was thinking to default to Deny before, but ask for explicitly specified sounds good

Reply:
Changed the types as proposed, added default for v1alpha1 to `Deny`.